### PR TITLE
Refactor `F.batch_normalization` and ChainerMN backend implementations

### LINKED ISSUE #6533: Refactoring of Multi Node Batch Normalization in ChainerMN
There are many parts where Multi Node Batch Normalization in ChainerMN overlaps with BatchNormalization in Chainer.
So, we should refactor multi node batch normalization.

ToDo:

- [ ] add args for comm and backend to BatchNormalization in functions of Chainer
- [ ] use BatchNormalization in functions of Chainer in Multi Node Batch Normalization.

### PR DESCRIPTION
**Pull request by belldandyxtq**
This PR refactors the `batch_normalization` in `chainermn`.
The motivation is to merge the `batch_normalization` code in `chainermn` into the `batch_normalization` in `chainer`, which can reduce the code redundancy, and enables features in `batch_normalization` in `chainer` into `chainermn`.

This PR resolves #6533

### CODE REVIEW

File: chainer/functions/normalization/batch_normalization.py [hunk 1]
```diff
@@ -15,6 +17,227 @@
 import chainerx
 
 
+class _BatchNormalizationBackend(six.with_metaclass(ABCMeta)):
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: As a convention, Chainer does not use `abc` for abstract classes, but just inherit from `object`.
As we have dropped Python 2, inheriting from `object` is not necessary either.
How about simply `class _BatchNormalizationBackend:`?

File: chainer/functions/normalization/batch_normalization.py [hunk 2]
```diff
@@ -15,6 +17,227 @@
...
+
+    @abstractmethod
+    def backward(self, axis, gamma, gy, x, xp,
+                 expander, mean, inv_std, eps, var, mode):
+        raise NotImplementedError()
+
+
+class _GeneralBatchNormalizationBackend(_BatchNormalizationBackend):
+
+    def __init__(self, mode=None):
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: I think `_BNMode` can now be integrated into backend selectors.
How do you think?

Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: You mean putting fields in `_BNMode` to corresponding `backend`?
Since `_BNMode` is also used in `FixedBatchNormalizationGrad`, removing it means we also need to refactor `FixedBatchNormalizationGrad`, which is not included in this patch.
Do you think we should do it in this PR? or another PR.

File: chainermn/functions/batch_normalization.py [hunk 3]
```diff
@@ -142,192 +121,30 @@ def get_communication_backend(comm, communication_backend='auto'):
...
-        self.mean_cache = None
-        self.decay = decay
-
-        selected_communication_backend = \
-            get_communication_backend(comm, communication_backend)
+        self.communication_backend = communication_backend
 
-        if selected_communication_backend == 'nccl':
-            self._backend = _NcclBackend(comm)
+    def _bn_backend_selector(self, xp):
```
Comment (niboshi) [inline chainermn/functions/batch_normalization.py]: How about passing a "selector" object to `BatchNormalization.__init__`, instead of subclassing it?

File: chainer/functions/normalization/batch_normalization.py (hunk 1 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: Addressed in 56f3d93

File: chainer/functions/normalization/batch_normalization.py (hunk 2 above)
Comment (niboshi) [reply chainer/functions/normalization/batch_normalization.py]: >You mean putting fields in `_BNMode` to corresponding `backend`?

No, I meant "selectors", not "backends". The role of `_BNMode` is overlapping with selectors.

OK, I understand it's not easy to simply eliminate `_BNMode`. But I don't think it doesn't have to be passed to backends. Their only usage is `is_for_conv2d` and `get_cudnn_mode` in `_CudnnBatchNormalizationBackend`. How about passing those two values to `_CudnnBatchNormalizationBackend.__init__`, instead of passing the `mode`?

File: chainermn/functions/batch_normalization.py (hunk 3 above)
Comment (belldandyxtq) [reply chainermn/functions/batch_normalization.py]: I think it would be difficult to make `selector` a separate function, as the parameters used to select the backend are different for backend in `chainer`: (general, iDeep, cudnn) and backends in `chainermn`: (mpi, nccl). As shown here we need to access `self.comm` to create the backends, which only exists in `chainermn`. Currently, with as a member of the `BatchNormalization`, `selector` can access all the members via `self`, which hide such difference. If we make `selector` a function, then we need to call it with all the parameters needed, i.e. having `self.comm` even in BN in `chainer`. 
Another option is to pass the `BatchNormalization` object itself to the `selector`, but it kinda creates a loop like:
`self.bn_backend = self.bn_selector(self)`

After discussing with @shu65, we adopted this approach.

Comment (niboshi) [reply chainermn/functions/batch_normalization.py]: Thanks.
`comm` is available at the moment of `MultiNodeBatchNormalizationFunction.__init__()` call.
I think you could write this way:

```py
def _create_selector(comm):
    def selector(other parameters...):
        if ...:
            return _NcclBackend(comm)
        ...
    return selector

func = BatchNormalization(..., _create_selector(comm))
```

`_create_selector` could also be replaced with a callable class.

It might look tricky, but I think the whole architecture is simpler because we won't need inheritance.

File: chainer/functions/normalization/batch_normalization.py (hunk 2 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: Addressed in 918119f

**Comment by shu65** [comment]
I'm performing the unittest for chainermn

File: chainermn/functions/batch_normalization.py (hunk 3 above)
Comment (belldandyxtq) [reply chainermn/functions/batch_normalization.py]: I made a draft as 36b1549, can you see if you like it?

File: chainer/functions/normalization/batch_normalization.py [hunk 4]
```diff
@@ -15,6 +15,233 @@
...
 
 
+class _BatchNormalizationBackend:
+
+    def forward(self, axis, gamma, x, xp, expander,
+                beta, eps, decay, running_mean, running_var):
+        raise NotImplementedError()
+
+    def backward(self, axis, gamma, gy, x, xp,
+                 expander, mean, inv_std, eps, var, mode):
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: I think `mode` argument is not used?

File: chainer/functions/normalization/batch_normalization.py [hunk 5]
```diff
@@ -152,205 +395,69 @@ def forward(self, inputs):
...
-
-                self.running_mean *= self.decay
-                self.running_mean += (1 - self.decay) * self.mean
-                self.running_var *= self.decay
-                self.running_var += (1 - self.decay) * adjust * var
-
-                if xp is chainerx:
-                    self.running_mean = backend.to_chx(self.running_mean)
-                    self.running_var = backend.to_chx(self.running_var)
+        self.mode = _BNMode(x, gamma, self.key_axis)
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: I think it's not necessary to keep `self.mode` for backward.

File: chainer/functions/normalization/batch_normalization.py [hunk 6]
```diff
@@ -152,205 +395,69 @@ def forward(self, inputs):
...
+        self.bn_backend = self.bn_backend_selector(xp, self.mode)
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: Rather than passing `self.mode` to the selector, how about passing `self` and `inputs`?
`_BNMode` is a highly internal entity, not appropriate for interfacing.

File: chainer/functions/normalization/batch_normalization.py (hunk 5 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: addressed in 8be6e49

File: chainer/functions/normalization/batch_normalization.py (hunk 4 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: addressed in 8be6e49

File: chainer/functions/normalization/batch_normalization.py (hunk 6 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: I am not sure I understand. Currently, we need `can_use_cudnn`, `can_use_ideep`, `is_for_conv2d`, `get_cudnn_mode` from `mode`. Do you mean we put the code to calculate these four into the selector?
https://github.com/chainer/chainer/blob/5080ad1c4faa0caf8cc14fea1a07dde30e066fe7/chainer/functions/normalization/batch_normalization.py#L265

Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: As the `mode` is removed from member in 8be6e49. Passing `self` cannot have access the `mode`

Comment (niboshi) [reply chainer/functions/normalization/batch_normalization.py]: `_BNMode` can be created inside `_bn_backend_selector` from `self` and `inputs`, as it's not used in `multinode_bn_backend_selector`.

Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: I see, I thought you mean we should not call anything related to `_BNMode` from outside of `BatchNormalization`.

Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: I made a draft in d218209. Actually, we only need the `key_axis` from `BatchNormalization` to create `_BNMode`. But I thought what would be too limited, what do you think.

File: chainer/functions/normalization/batch_normalization.py [hunk 7]
```diff
@@ -65,6 +309,8 @@ def __init__(self, eps=2e-5, mean=None, var=None, decay=0.9, axis=None):
             raise RuntimeError('axis must be int, tuple of int or None')
         self.axis = axis
 
+        self.bn_backend_selector = backend_selector
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: How about `self._backend_selector`?
I think `bn` is redundant because this class is `BatchNormalization`.

File: chainer/functions/normalization/batch_normalization.py [hunk 8]
```diff
@@ -152,205 +398,66 @@ def forward(self, inputs):
...
-            beta = beta[expander].astype(interm_dtype, copy=False)
-            self.mean = x.mean(axis=self.axis, dtype=interm_dtype)
-            var = x.var(axis=self.axis, dtype=interm_dtype)
-            if xp is numpy:
-                self.inv_std = numpy.reciprocal(numpy.sqrt(
-                    var + self.eps, dtype=interm_dtype))
-            else:
-                self.inv_std = cuda.cupyx.rsqrt(
-                    var + self.eps, dtype=interm_dtype)
+        self.bn_backend = self.bn_backend_selector(self, inputs)
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: Ditto. How about `self._backend`?

File: chainer/functions/normalization/batch_normalization.py [hunk 9]
```diff
@@ -15,6 +15,233 @@
...
+    def __init__(self):
+        self.mean = None
+        self.inv_std = None
+        self.var = None
+        self.running_mean = None
+        self.running_var = None
+
+    def forward(self, axis, gamma, x, xp, expander,
+                beta, eps, decay, running_mean, running_var):
+        self.running_mean = running_mean
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: Rather than interfacing using attributes in `self`, I think it's simpler to include all the input/output requirements in arguments and return values.
So, how about returning them (`mean`, `inv_std`, `var`, `running_mean`, `running_var`) along with `y` instead of assigning attributes?

File: chainer/functions/normalization/batch_normalization.py [hunk 10]
```diff
@@ -15,6 +15,233 @@
...
+                    'update_mean_var')(self.mean, var, decay, adjust,
+                                       self.running_mean, self.running_var)
+
+            if xp is chainerx:
+                self.running_mean = backend.to_chx(self.running_mean)
+                self.running_var = backend.to_chx(self.running_var)
+
+        return y,
+
+    def _get_mean_and_var(self, axis, gamma, x, xp, interm_dtype):
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: As this method is overridden, it's not internal. Please remove the prefix `_`.

File: chainer/functions/normalization/batch_normalization.py [hunk 11]
```diff
@@ -15,6 +15,233 @@
...
+        mean = x.mean(axis=axis, dtype=interm_dtype)
+        var = x.var(axis=axis, dtype=interm_dtype)
+        return mean, var
+
+    def _get_ggamma_and_gbeta(self, axis, gamma, gy, x_hat, xp):
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: Ditto. Please remove the prefix `_`.

File: chainer/functions/normalization/batch_normalization.py (hunk 6 above)
Comment (niboshi) [reply chainer/functions/normalization/batch_normalization.py]: Currently only `key_axis` is required because of cuDNN's specification. I don't think it's a good idea to encode that into the interface.
We could pass all the possibly-useful attributes, but I think passing `BatchNormalization` is an easier option for now.

File: chainer/functions/normalization/batch_normalization.py (hunk 7 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: I used `bn` because there is already a `backend` there. 
https://github.com/chainer/chainer/blob/d2182099fcc59ac11daf358b496651730aed933c/chainer/functions/normalization/batch_normalization.py#L7
Might cause confusion

File: chainer/functions/normalization/batch_normalization.py [hunk 12]
```diff
@@ -34,12 +261,29 @@ def _compute_key_axis(x_ndim, gamma_ndim=1, axis=None):
     return key_axis
 
 
+def _bn_backend_selector(batch_norm_obj, inputs):
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: How about `func` instead of `obj`?

File: chainermn/functions/batch_normalization.py [hunk 13]
```diff
@@ -142,192 +119,13 @@ def get_communication_backend(comm, communication_backend='auto'):
...
 
 
-class MultiNodeBatchNormalizationFunction(function.Function):
-
-    def __init__(self, comm, eps=2e-5, mean=None, var=None, decay=0.9,
-                 communication_backend='auto'):
-        chainer.utils.experimental(
-            'chainermn.functions.MultiNodeBatchNormalizationFunction')
-
+class multinode_bn_backend_selector:
```
Comment (niboshi) [inline chainermn/functions/batch_normalization.py]: Please use CamelCase for classes.

File: chainermn/links/batch_normalization.py [hunk 14]
```diff
@@ -10,7 +10,9 @@
 from chainermn.functions.batch_normalization import \
     get_communication_backend
 from chainermn.functions.batch_normalization import \
-    MultiNodeBatchNormalizationFunction
+    multinode_bn_backend_selector
+from chainer.functions.normalization.batch_normalization \
+    import BatchNormalization
```
Comment (niboshi) [inline chainermn/links/batch_normalization.py]: Lets' avoid importing objects, as per [the OpenStack Style Guideline](https://docs.openstack.org/hacking/latest/user/hacking.html#imports).

File: chainer/functions/normalization/batch_normalization.py (hunk 9 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: addressed in 84628a4

File: chainer/functions/normalization/batch_normalization.py (hunk 11 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: Addressed in 4d0e801

File: chainer/functions/normalization/batch_normalization.py (hunk 10 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: Addressed in 4d0e801

File: chainermn/functions/batch_normalization.py (hunk 13 above)
Comment (belldandyxtq) [reply chainermn/functions/batch_normalization.py]: Addressed in 4c2fe6c

File: chainermn/links/batch_normalization.py (hunk 14 above)
Comment (belldandyxtq) [reply chainermn/links/batch_normalization.py]: Addressed in 4c2fe6c

File: chainer/functions/normalization/batch_normalization.py (hunk 12 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: Addressed in 4c2fe6c

File: chainer/functions/normalization/batch_normalization.py (hunk 6 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: Thanks. Then I think d218209 has addressed the issue

File: chainer/functions/normalization/batch_normalization.py (hunk 7 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: Similarly, we have `communication_backend` in `chainermn`

Comment (niboshi) [reply chainer/functions/normalization/batch_normalization.py]: Hmm, I see. Then, how about using a different word than `backend`?

How about `impl`? This word is used in pooling implementation (`_CudnnPollingNDImpl` in `chainer/functions/pooling/pooilng_nd.py`). This class is semantically similar to `_BatchNormalizationBackend` class in this PR.

More specifically, my suggestion is:
* `_BatchNormalizationBackend` -> `_BatchNormalizationImpl`
* `BatchNormalization.bn_backend_selector` -> `BatchNormalization._impl_selector`
* `BatchNormalization(Grad).bn_backend` -> `BatchNormalization(Grad)._impl`

Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: Addressed in 4fae569

File: chainer/functions/normalization/batch_normalization.py (hunk 8 above)
Comment (belldandyxtq) [reply chainer/functions/normalization/batch_normalization.py]: Addressed in 4fae569

**Comment by niboshi** [comment]
Thanks, I think this PR makes the code much organized than before.
Jenkins, test this please

**Comment by niboshi** [comment]
The test failure does not seem related.
